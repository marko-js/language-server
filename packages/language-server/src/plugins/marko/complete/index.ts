import { NodeType } from "@marko/language-tools";
import { CompletionItem } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export function provideCompletions(
  doc: MarkoVirtualCode,
  offset: number,
): CompletionItem[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc, offset);
    case NodeType.Import:
    case NodeType.Static:
      return Import(node, doc);
    case NodeType.Tag:
      return Tag(node, doc, offset);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
