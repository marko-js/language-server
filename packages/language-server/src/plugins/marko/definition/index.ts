import { NodeType } from "@marko/language-tools";
import { LocationLink } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export function provideDefinitions(
  doc: MarkoVirtualCode,
  offset: number,
): LocationLink[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
