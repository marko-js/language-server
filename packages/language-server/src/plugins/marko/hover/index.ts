import { NodeType } from "@marko/language-tools";
import { Hover } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { OpenTagName } from "./OpenTagName";

export function provideHover(
  doc: MarkoVirtualCode,
  offset: number,
): Hover | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
