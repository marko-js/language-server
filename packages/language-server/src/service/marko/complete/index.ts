import type { CompletionItem, CompletionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { AttrValue } from "./AttrValue";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export type CompletionResult = Result<CompletionItem[]>;
export interface CompletionMeta<N = unknown> {
  file: MarkoFile;
  params: CompletionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: CompletionMeta<any>) => CompletionResult
> = {
  Tag,
  OpenTagName,
  AttrName,
  AttrValue,
  Import,
};

export const doComplete: Plugin["doComplete"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return {
    items:
      (await handlers[NodeType[node.type]]?.({
        file,
        params,
        offset,
        node,
      })) || [],
    isIncomplete: true,
  };
};
