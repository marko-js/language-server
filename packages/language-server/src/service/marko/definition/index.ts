import type { DefinitionLink, DefinitionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export type DefinitionResult = Result<DefinitionLink[]>;
export interface DefinitionMeta<N = unknown> {
  file: MarkoFile;
  params: DefinitionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: DefinitionMeta<any>) => DefinitionResult
> = {
  OpenTagName,
  AttrName,
};

export const findDefinition: Plugin["findDefinition"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return (
    (await handlers[NodeType[node.type]]?.({
      file,
      params,
      offset,
      node,
    })) || []
  );
};
