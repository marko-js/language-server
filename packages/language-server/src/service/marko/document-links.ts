import type { DocumentLink } from "vscode-languageserver";
import { type Node, NodeType } from "@marko/language-tools";
import { MarkoFile, processDoc } from "../../utils/file";

import resolveUrl from "../../utils/resolve-url";
import type { Plugin } from "../types";

import isDocumentLinkAttr from "./util/is-document-link-attr";

const importTagReg = /(['"])<((?:[^\1\\>]+|\\.)*)>?\1/g;

export const findDocumentLinks: Plugin["findDocumentLinks"] = async (doc) => {
  return processDoc(doc, extractDocumentLinks);
};

/**
 * Iterate over the Marko CST and extract all the file links in the document.
 */
function extractDocumentLinks({
  uri,
  scheme,
  parsed,
  code,
  project: { lookup },
}: MarkoFile): DocumentLink[] {
  if (scheme !== "file") {
    return [];
  }

  const links: DocumentLink[] = [];
  const { program, read } = parsed;
  const visit = (node: Node.ChildNode) => {
    switch (node.type) {
      case NodeType.AttrTag:
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }
        break;
      case NodeType.Tag:
        if (node.attrs && node.nameText) {
          for (const attr of node.attrs) {
            if (isDocumentLinkAttr(code, node, attr)) {
              const resolved = resolveUrl(
                read(attr.value.value).slice(1, -1),
                uri
              );
              if (resolved) {
                links.push({
                  range: parsed.locationAt(attr.value.value),
                  target: resolveUrl(read(attr.value.value).slice(1, -1), uri),
                });
              }
            }
          }
        }
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }

        break;
    }
  };

  for (const node of program.static) {
    // check for import statement (this currently only support the tag import shorthand).
    if (node.type === NodeType.Import) {
      importTagReg.lastIndex = 0;
      const value = parsed.read(node);
      const match = importTagReg.exec(value);
      if (match) {
        const [{ length }, , tagName] = match;
        const tagDef = lookup.getTag(tagName);
        const fileForTag = tagDef && (tagDef.template || tagDef.renderer);

        if (fileForTag) {
          links.push({
            range: parsed.locationAt({
              start: node.start + match.index,
              end: node.start + match.index + length,
            }),
            target: fileForTag,
          });
        }
      }
    }
  }

  for (const node of program.body) {
    visit(node);
  }

  return links;
}
