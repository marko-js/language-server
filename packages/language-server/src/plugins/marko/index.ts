import {
  LanguageServicePlugin,
  LanguageServicePluginInstance,
} from "@volar/language-service";
import { TextDocument } from "vscode-languageserver-textdocument";
import { URI } from "vscode-uri";

import { MarkoVirtualCode } from "../../language";
import { provideCompletions } from "./complete";
import { provideDefinitions } from "./definition";
import { provideHover } from "./hover";
import { provideValidations } from "./validate";
// import { provideDocumentSymbols } from "./document-symbols";

export const create = (
  _: typeof import("typescript"),
): LanguageServicePlugin => {
  return {
    name: "marko",
    capabilities: {
      hoverProvider: true,
      definitionProvider: true,
      diagnosticProvider: {
        interFileDependencies: false,
        workspaceDiagnostics: false,
      },
      completionProvider: {
        triggerCharacters: [
          ".",
          ":",
          "<",
          ">",
          "@",
          "/",
          '"',
          "'",
          "`",
          " ",
          "=",
          "*",
          "#",
          "$",
          "+",
          "^",
          "(",
          "[",
          "-",
        ],
      },
    },
    create(context): LanguageServicePluginInstance {
      return {
        // TODO: Is this necessary?
        // provideDocumentSymbols(document, token) {
        //   if (token.isCancellationRequested) return;
        //   return worker(document, (virtualCode) => {
        //     return provideDocumentSymbols(virtualCode);
        //   });
        // },
        provideDefinition(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideDefinitions(virtualCode, offset);
          });
        },
        provideDiagnostics(document, token) {
          if (token.isCancellationRequested) return;
          return worker(document, async (virtualCode) => {
            return await provideValidations(virtualCode);
          });
        },
        provideHover(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideHover(virtualCode, offset);
          });
        },
        provideCompletionItems(document, position, _, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            const completions = provideCompletions(virtualCode, offset);

            if (completions) {
              return {
                isIncomplete: false,
                items: completions.map((it) => {
                  it.data = { source: "marko" };
                  return it;
                }),
              };
            }

            return {
              items: [],
              isIncomplete: true,
            };
          });
        },
      };

      function worker<T>(
        document: TextDocument,
        callback: (markoDocument: MarkoVirtualCode) => T,
      ): T | undefined {
        const decoded = context.decodeEmbeddedDocumentUri(
          URI.parse(document.uri),
        );
        const sourceScript =
          decoded && context.language.scripts.get(decoded[0]);
        const virtualCode =
          decoded && sourceScript?.generated?.embeddedCodes.get(decoded[1]);
        if (!(virtualCode instanceof MarkoVirtualCode)) return;

        return callback(virtualCode);
      }
    },
  };
};
