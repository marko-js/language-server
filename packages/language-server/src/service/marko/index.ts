import type { Plugin } from "../types";

import { doComplete } from "./complete";
import { doValidate } from "./validate";
import { doHover } from "./hover";
import { findDefinition } from "./definition";
import { findDocumentLinks } from "./document-links";
import { findDocumentSymbols } from "./document-symbols";
import { format } from "./format";

export default {
  doComplete,
  doValidate,
  doHover,
  findDefinition,
  findDocumentLinks,
  findDocumentSymbols,
  format,
} as Plugin;
