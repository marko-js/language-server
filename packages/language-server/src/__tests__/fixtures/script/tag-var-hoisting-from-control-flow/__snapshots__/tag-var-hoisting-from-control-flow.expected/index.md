## Hovers
### Ln 7, Col 7
```marko
   5 |   <for|d| of=b>
   6 |     <const/c=a.filter((e) => e.length)/>
>  7 |     ${d}
     |       ^ (parameter) d: "slice" | "dice"
   8 | //    ^?
   9 |   </for>
  10 | </if>
```

### Ln 14, Col 3
```marko
  12 |
  13 | -- ${() => {
> 14 |   a;
     |   ^ const a: readonly ["apples", "oranges"] | undefined
  15 | //^?
  16 |   b;
  17 | //^?
```

### Ln 16, Col 3
```marko
  14 |   a;
  15 | //^?
> 16 |   b;
     |   ^ const b: readonly ["slice", "dice"] | undefined
  17 | //^?
  18 |   c;
  19 | //^?
```

### Ln 18, Col 3
```marko
  16 |   b;
  17 | //^?
> 18 |   c;
     |   ^ const c: ("apples" | "oranges")[] | undefined
  19 | //^?
  20 | }}
  21 |
```

