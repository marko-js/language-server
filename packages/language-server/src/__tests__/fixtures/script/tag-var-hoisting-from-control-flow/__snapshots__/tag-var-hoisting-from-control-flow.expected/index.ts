export interface Input {}
abstract class Component extends Marko.Component<Input> {}
export { type Component };
(function (this: void) {
  const input = Marko._.any as Input;
  const component = Marko._.any as Component;
  const state = Marko._.state(component);
  const out = Marko._.out;
  const $signal = Marko._.any as AbortSignal;
  const $global = Marko._.getGlobal(
    // @ts-expect-error We expect the compiler to error because we are checking if the MarkoRun.Context is defined.
    (Marko._.error, Marko._.any as MarkoRun.Context),
  );
  const __marko_internal_rendered_1 = (() => {
    if (true) {
      const __marko_internal_tag_1 = Marko._.resolveTemplate(
        import("../../../components/const/index.marko"),
      );
      const __marko_internal_rendered_3 = Marko._.renderTemplate(
        __marko_internal_tag_1,
      )()()({
        value: ["apples", "oranges"] as const,
      });
      const a = __marko_internal_rendered_3.return.value;
      const __marko_internal_tag_2 = Marko._.resolveTemplate(
        import("../../../components/const/index.marko"),
      );
      const __marko_internal_rendered_4 = Marko._.renderTemplate(
        __marko_internal_tag_2,
      )()()({
        value: ["slice", "dice"] as const,
      });
      const b = __marko_internal_rendered_4.return.value;
      const __marko_internal_rendered_2 = Marko._.forOfTag(
        {
          /*for*/ of: b,
        },
        (d) => {
          const __marko_internal_tag_3 = Marko._.resolveTemplate(
            import("../../../components/const/index.marko"),
          );
          const __marko_internal_rendered_5 = Marko._.renderTemplate(
            __marko_internal_tag_3,
          )()()({
            value: a.filter((e) => e.length),
          });
          const c = __marko_internal_rendered_5.return.value;
          d;
          return new (class MarkoReturn<Return = void> {
            [Marko._.scope] = { c };
            declare return: Return;
            constructor(_?: Return) {}
          })();
        },
      );
      return {
        scope: { ...Marko._.readScope(__marko_internal_rendered_2), a, b },
      };
    } else {
      return undefined;
    }
  })();
  //    ^?
  () => {
    a;
    //^?
    b;
    //^?
    c;
    //^?
  };
  const { a, b, c } = Marko._.readScope(__marko_internal_rendered_1);
  Marko._.noop({ a, b, c, component, state, out, input, $global, $signal });
  return;
})();
export default new (class Template extends Marko._.Template<{
  render(
    input: Marko.TemplateInput<Input>,
    stream?: {
      write: (chunk: string) => void;
      end: (chunk?: string) => void;
    },
  ): Marko.Out<Component>;

  render(
    input: Marko.TemplateInput<Input>,
    cb?: (err: Error | null, result: Marko.RenderResult<Component>) => void,
  ): Marko.Out<Component>;

  renderSync(input: Marko.TemplateInput<Input>): Marko.RenderResult<Component>;

  renderToString(input: Marko.TemplateInput<Input>): string;

  stream(
    input: Marko.TemplateInput<Input>,
  ): ReadableStream<string> & NodeJS.ReadableStream;

  mount(
    input: Marko.TemplateInput<Input>,
    reference: Node,
    position?: "afterbegin" | "afterend" | "beforebegin" | "beforeend",
  ): Marko.MountedTemplate<typeof input>;

  api: "class";
  _(): () => <__marko_internal_input extends unknown>(
    input: Marko.Directives &
      Input &
      Marko._.Relate<__marko_internal_input, Marko.Directives & Input>,
  ) => Marko._.ReturnWithScope<__marko_internal_input, void>;
}> {})();
