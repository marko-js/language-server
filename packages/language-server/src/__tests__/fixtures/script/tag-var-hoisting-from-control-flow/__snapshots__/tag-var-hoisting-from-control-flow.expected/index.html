<div>
  
  

  <div>
    
    placeholder

  </div>
</div>placeholder